File: python/routes/tasks.py
from flask import Blueprint, jsonify, request

from models.task import (
    DUE_DATE_KEY,
    SORT_OPTIONS,
    create_task,
    delete_task,
    parse_due_date,
    task_list_view,
    task_view,
    update_task,
)

tasks_bp = Blueprint("tasks", __name__, url_prefix="/tasks")

SORT_PARAM = "sort"
INVALID_DUE_DATE_ERROR = "dueDate must be a valid ISO 8601 datetime"
INVALID_SORT_ERROR = "sort must be one of: " + ", ".join(SORT_OPTIONS)


@tasks_bp.get("")
def list_tasks():
    sort = request.args.get(SORT_PARAM)
    if sort is not None and sort not in SORT_OPTIONS:
        return jsonify({"error": INVALID_SORT_ERROR}), 400
    return jsonify(task_list_view(sort))


@tasks_bp.post("")
def add_task():
    body = request.get_json(silent=True) or {}
    title = body.get("title")
    if not title:
        return jsonify({"error": "title is required"}), 400
    try:
        due_date = parse_due_date(body.get(DUE_DATE_KEY))
    except ValueError:
        return jsonify({"error": INVALID_DUE_DATE_ERROR}), 400
    return jsonify(task_view(create_task(title, due_date))), 201


@tasks_bp.patch("/<int:task_id>")
def patch_task(task_id):
    completed = (request.get_json(silent=True) or {}).get("completed")
    task = update_task(task_id, bool(completed))
    if task is None:
        return jsonify({"error": "task not found"}), 404
    return jsonify(task_view(task))


@tasks_bp.delete("/<int:task_id>")
def remove_task(task_id):
    if not delete_task(task_id):
        return jsonify({"error": "task not found"}), 404
    return "", 204

File: Node/routes/tasks.js
const express = require('express');
const {
  createTask,
  getTasks,
  getTask,
  updateTask,
  setDueDate,
  deleteTask,
} = require('../models/task');

const router = express.Router();

router.get('/', (req, res) => {
  res.json(getTasks());
});

router.post('/', (req, res) => {
  const { title, due_date } = req.body;
  if (!title) {
    return res.status(400).json({ error: 'title is required' });
  }
  // due_date is optional; absent -> null. Stored value is normalized to IST.
  const task = createTask(title, due_date ?? null);
  res.status(201).json(task);
});

router.patch('/:id', (req, res) => {
  const id = Number(req.params.id);
  const task = getTask(id);
  if (!task) {
    return res.status(404).json({ error: 'task not found' });
  }

  if ('due_date' in req.body) {
    // Immutable once set: reject any attempt to supply due_date when one exists.
    if (task.due_date !== null && task.due_date !== undefined) {
      return res.status(400).json({ error: 'due_date cannot be changed once set' });
    }
    setDueDate(id, req.body.due_date);
  }
  if ('completed' in req.body) {
    updateTask(id, { completed: req.body.completed });
  }

  res.json(getTask(id));
});

router.delete('/:id', (req, res) => {
  const id = Number(req.params.id);
  const deleted = deleteTask(id);
  if (!deleted) {
    return res.status(404).json({ error: 'task not found' });
  }
  res.status(204).send();
});

module.exports = router;
